Compile failures under Arduino 1.6.8+ (with fix)

I suddenly found that compilation of even an empty sketch would fail either during compilation or linking of the Arduino core libraries under Arduino 1.6.8 (and the 1.6.9 nightlies). Looking at the history of the Arduino repo, it looks like everything back to Arduino 1.6.6 is probably also affected. (FastLED under any version is also affected.)

Adding `CXXFLAGS = --std=gnu++11 -fno-threadsafe-statics`
was necessary to get anything to build (the Arduino IDE uses these flags, which is how I found them).

Not sure what changed, but it looks like those two flags need to be added.

Example compile errors:
```
/Applications/Arduino.app/Contents/Java/hardware/tools/avr/bin/avr-g++ -MMD -c -mmcu=atmega32u4 -DF_CPU=16000000L -DARDUINO=nig -DARDUINO_ARCH_AVR -D__PROG_TYPES_COMPAT__ -I/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino -I/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/variants/leonardo     -I../libraries/FastLED -Wall -ffunction-sections -fdata-sections -Os -DUSB_VID=0x2341 -DUSB_PID=0x8036  -fno-exceptions  /Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp -o build-leonardo/core/USBCore.cpp.o
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp: In function 'int USB_RecvControl(void*, int)':
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp:431:2: warning: 'auto' changes meaning in C++11; please remove it [-Wc++0x-compat]
  auto length = len;
  ^
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp:431:7: error: 'length' does not name a type
  auto length = len;
       ^
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp:432:8: error: 'length' was not declared in this scope
  while(length)
        ^
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp:436:3: warning: 'auto' changes meaning in C++11; please remove it [-Wc++0x-compat]
   auto recvLength = length;
   ^
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp:436:8: error: 'recvLength' does not name a type
   auto recvLength = length;
        ^
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp:437:6: error: 'recvLength' was not declared in this scope
   if(recvLength > 64){
      ^
/Applications/Arduino.app/Contents/Java/hardware/arduino/avr/cores/arduino/USBCore.cpp:443:31: error: 'recvLength' was not declared in this scope
   Recv((u8*)d + len - length, recvLength);
                               ^
make: *** [build-leonardo/core/USBCore.cpp.o] Error 1
```

```
/Applications/Arduino.app/Contents/Java/hardware/tools/avr/bin/avr-gcc -mmcu=atmega32u4 -Wl,--gc-sections -Os -o build-leonardo/neopixel2.elf build-leonardo/neopixel2.ino.o build-leonardo/libcore.a  -lc -lm 
build-leonardo/neopixel2.ino.o: In function `setup':
neopixel2.ino:(.text.setup+0xc): undefined reference to `__cxa_guard_acquire'
neopixel2.ino:(.text.setup+0x9e): undefined reference to `__cxa_guard_release'
build-leonardo/libcore.a(PluggableUSB.cpp.o): In function `PluggableUSB()':
PluggableUSB.cpp:(.text._Z12PluggableUSBv+0xc): undefined reference to `__cxa_guard_acquire'
PluggableUSB.cpp:(.text._Z12PluggableUSBv+0x2c): undefined reference to `__cxa_guard_release'
collect2: error: ld returned 1 exit status
make: *** [build-leonardo/neopixel2.elf] Error 1
```

I just ran into this yesterday.  I am on OS X.  No issues on empty loop and setup functions, but if I had a line for FastLED, it would error:

FastLED line in setup():

      FastLED.addLeds<LED_TYPE,DATA_PIN,COLOR_ORDER>(leds, NUM_LEDS).setCorrection(TypicalLEDStrip);

Error messages:

    build-uno/FastLED_DemoReel100.cpp.o: In function `setup':
    FastLED_DemoReel100.cpp:(.text.setup+0x2a): undefined reference to `__cxa_guard_acquire'
    FastLED_DemoReel100.cpp:(.text.setup+0xbc): undefined reference to `__cxa_guard_release'
    collect2: error: ld returned 1 exit status
    make: *** [build-uno/FastLED_DemoReel100.elf] Error 1

If I added the line listed above to my Makefile, it worked.  As far as I have gotten is building.  Have sent anything to the board yet.  Let me know if any additional info would be helpful.

--- a/Arduino.mk
+++ b/Arduino.mk
@@ -472,8 +472,12 @@ ifndef AVR_TOOLS_PATH
     AVR_TOOLS_PATH    = $(AVR_TOOLS_DIR)/bin
 endif
 
-ARDUINO_LIB_PATH  = $(ARDUINO_DIR)/libraries
-$(call show_config_variable,ARDUINO_LIB_PATH,[COMPUTED],(from ARDUINO_DIR))
+ifndef ARDUINO_LIB_PATH
+    ARDUINO_LIB_PATH = $(ARDUINO_DIR)/libraries
+    $(call show_config_variable,ARDUINO_LIB_PATH,[COMPUTED],(from ARDUINO_DIR))
+else
+    $(call show_config_variable,ARDUINO_LIB_PATH,[USER])
+endif
 
 # 1.5.x platform dependent libs path
 ifndef ARDUINO_PLATFORM_LIB_PATH
@@ -1035,7 +1039,11 @@ else
 endif
 
 ifndef CXXFLAGS_STD
-    CXXFLAGS_STD      =
+    ifeq ($(shell expr $(ARDUINO_VERSION) '>' 150), 1)
+        CXXFLAGS_STD      = -std=gnu++11 -fno-threadsafe-statics
+    else
+        CXXFLAGS_STD      =
+    endif
     $(call show_config_variable,CXXFLAGS_STD,[DEFAULT])
 else
     $(call show_config_variable,CXXFLAGS_STD,[USER])

--- a/arduino-mk-vars.md
+++ b/arduino-mk-vars.md
@@ -929,7 +929,7 @@ CFLAGS_STD = = -std=gnu89
 
 Controls, *exclusively*, which C++ standard is to be used for compilation.
 
-Defaults to `undefined`
+Defaults to `undefined` on 1.0 or `-std=gnu++11 -fno-threadsafe-statics` on 1.5+
 
 Possible values:
 

